nuxeo-features/nuxeo-platform-oauth/src/main/java/org/nuxeo/ecm/platform/ui/web/auth/oauth2/NuxeoOAuth2Servlet.java

```diff
@@ -0,0 +1,246 @@
+            });
+
+            handleTokenResponse(token, response);
+        } else if ("refresh_token".equals(tokRequest.getGrantType())) {
+            OAuth2Error error = null;
+            if (isBlank(tokRequest.getClientId())) {
+                error = OAuth2Error.access_denied;
+            } else if (!clientRegistry.isValidClient(tokRequest.getClientId(), tokRequest.getClientSecret())) {
+                error = OAuth2Error.access_denied;
+            }
+
+            if (error != null) {
+                handleJsonError(error, request, response);
+                return;
+            }
+
+            NuxeoOAuth2Token refreshed = TransactionHelper.runInNewTransaction(
+                    () -> getTokenStore().refresh(tokRequest.getRefreshToken(), tokRequest.getClientId()));
+
+            if (refreshed == null) {
+                handleJsonError(OAuth2Error.invalid_request, request, response);
+            } else {
+                handleTokenResponse(refreshed, response);
+            }
+        } else {
+            handleJsonError(OAuth2Error.invalid_grant, request, response);
+        }
+    }
+
+    protected void handleTokenResponse(NuxeoOAuth2Token token, HttpServletResponse response) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+
+        response.setHeader("Content-Type", "application/json");
+        response.setStatus(200);
+        mapper.writeValue(response.getWriter(), token.toJsonObject());
+    }
+
+    protected void handleError(OAuth2Error error, HttpServletRequest request, HttpServletResponse response,
+            String redirectURI) throws IOException {
+        Map<String, String> params = new HashMap<>();
+        params.put("error", error.toString());
+        String state = request.getParameter("state");
+        if (isNotBlank(state)) {
+            params.put("state", state);
+        }
+
+        sendRedirect(response, redirectURI, params);
+    }
+
+    protected void handleJsonError(OAuth2Error error, HttpServletRequest request, HttpServletResponse response)
+            throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+
+        response.setHeader("Content-Type", "application/json");
+        response.setStatus(400);
+
+        Map<String, String> object = new HashMap<>();
+        object.put("error", error.toString());
+        mapper.writeValue(response.getWriter(), object);
+    }
+
+    protected void sendRedirect(HttpServletResponse response, String uri, Map<String, String> params)
+            throws IOException {
+        if (uri == null) {
+            uri = "http://dummyurl";
+        }
+
+        StringBuilder sb = new StringBuilder(uri);
+        if (params != null) {
+            if (!uri.contains("?")) {
+                sb.append("?");
+            } else {
+                sb.append("&");
+            }
+
+            for (String key : params.keySet()) {
+                sb.append(key).append("=").append(params.get(key)).append("&");
```

Comment:
This ad hoc code is wrong, it's missing some escaping and encoding. We should use `URIUtils.addParametersToURIQuery` instead.